airflow/providers/facebook/ads/hooks/ads.py

```diff
@@ -0,0 +1,150 @@
+
+    :param facebook_ads_conn_id: Airflow Facebook Ads connection ID
+    :type facebook_ads_conn_id: str
+    :param api_version: The version of Facebook API. Default to v6.0
+    :type api_version: str
+
+    :return: list of Facebook Ads AdsInsights object(s)
+    :rtype: list[AdsInsights]
+    """
+
+    def __init__(
+        self,
+        facebook_ads_conn_id: str = "facebook_ads_default",
+        api_version: str = "v6.0",
+    ) -> None:
+        super().__init__()
+        self.facebook_ads_conn_id = facebook_ads_conn_id
+        self.facebook_ads_config: Dict[str, Any] = {}
+        self.api_version = api_version
+        self.ad_account_id: str = ""
+        self.app_id: str = ""
+        self.app_secret: str = ""
+        self.access_token: str = ""
+        self.default_params = {
+            'level': 'ad',
+            'date_preset': 'yesterday',
+        }
+
+    def _get_service(self) -> FacebookAdsApi:
+        """ Returns Facebook Ads Client using a service account"""
+        self._get_config()
+        for key, val in self.facebook_ads_config.items():
+            setattr(self, key, val)
+
+        return FacebookAdsApi.init(app_id=self.app_id,
+                                   app_secret=self.app_secret,
+                                   access_token=self.access_token,
+                                   api_version=self.api_version)
+
+    def _get_config(self) -> None:
+        """
+        Gets Facebook ads connection from meta db and sets
+        facebook_ads_config attribute with returned config file
+        """
+        conn = self.get_connection(self.facebook_ads_conn_id)
+        if "facebook_ads_client" not in conn.extra_dejson:
+            raise AirflowException("facebook_ads_client not found in extra field")
+
+        self.facebook_ads_config = conn.extra_dejson["facebook_ads_client"]
+
+    def bulk_facebook_report(
+        self,
+        params: Optional[Dict[str, Any]] = None,
+        fields: Optional[List[str]] = None,
+    ) -> List[AdsInsights]:
+        """
+        Pulls data from the Facebook Ads API
+
+        :param fields: List of fields that is obtained from Facebook. Found in AdsInsights.Field class.
+            https://developers.facebook.com/docs/marketing-api/insights/parameters/v6.0
+        :type fields: List[str]
+        :param params: Parameters that determine the query for Facebook
+            https://developers.facebook.com/docs/marketing-api/insights/parameters/v6.0
+
+        :return: Facebook Ads API response, converted to Facebook Ads Row objects
+        :rtype: List[AdsInsights]
+        """
+        api = self._get_service()
+        params = params if params else self.default_params
+        ad_account = AdAccount(self.ad_account_id, api=api)
+        _async = ad_account.get_insights(params=params, fields=fields, is_async=True)
+        try:
+            while _async.api_get()["async_status"] != JOB_COMPLETED:
```

Comment:
I'm do not know facebook API, is there possibility that the job will complete with other status than `JOB_COMPLETED`? For example with `JOB_FAILED` or something. I know others APIs have such behavior

Reply:
Yep, have solved will, re commit